transport/grpc/integration_test.go

```diff
@@ -72,9 +83,78 @@ func TestGRPCBasic(t *testing.T) {
 	})
 }
 
+func TestTLSWithYARPC(t *testing.T) {
+	for _, test := range []struct {
+		clientValidity      time.Duration
+		serverValidity      time.Duration
+		expectedErrContains string
+		name                string
+	}{
+		{
+			clientValidity: 1 * time.Minute,
+			serverValidity: 1 * time.Minute,
+			name:           "valid certs both sides",
+		},
+		{
+			clientValidity:      1 * time.Minute,
+			serverValidity:      -1 * time.Minute,
+			expectedErrContains: "transport: authentication handshake failed: x509: certificate has expired or is not yet valid",
+			name:                "invalid server cert",
+		},
+		{
+			clientValidity:      -1 * time.Minute,
+			serverValidity:      1 * time.Minute,
+			expectedErrContains: "remote error: tls: bad certificate",
+			name:                "invalid client cert",
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			scenario := createTLSScenario(t, test.clientValidity, test.serverValidity)
+
+			serverCreds := credentials.NewTLS(&tls.Config{
+				Certificates: []tls.Certificate{{
+					Certificate: [][]byte{scenario.ServerCert.Raw},
+					Leaf:        scenario.ServerCert,
+					PrivateKey:  scenario.ServerKey,
+				}},
+				ClientAuth: tls.RequireAndVerifyClientCert,
+				ClientCAs:  scenario.CAs,
+			})
+			inboundTLSOpt := InboundCredentials(serverCreds)
+
+			clientCreds := credentials.NewTLS(&tls.Config{
+				Certificates: []tls.Certificate{{
+					Certificate: [][]byte{scenario.ClientCert.Raw},
+					Leaf:        scenario.ClientCert,
+					PrivateKey:  scenario.ClientKey,
+				}},
+				RootCAs: scenario.CAs,
+			})
+			dialTLSOpt := DialerCredentials(clientCreds)
+
+			doWithTestEnv(t, nil, []InboundOption{inboundTLSOpt}, nil, []DialOption{dialTLSOpt}, func(t *testing.T, e *testEnv) {
+				err := e.SetValueYARPC(context.Background(), "foo", "bar")
+				if test.expectedErrContains == "" {
+					assert.NoError(t, err)
+				} else {
+					assert.Contains(t, err.Error(), test.expectedErrContains)
+				}
+
+				err = e.SetValueGRPC(context.Background(), "foo", "bar")
+				if test.expectedErrContains == "" {
+					assert.NoError(t, err)
+				} else {
+					assert.Contains(t, err.Error(), test.expectedErrContains)
+				}
+			})
+		})
+
+	}
+}
+
 func TestYARPCWellKnownError(t *testing.T) {
 	t.Parallel()
-	doWithTestEnv(t, nil, nil, nil, func(t *testing.T, e *testEnv) {
+	doWithTestEnv(t, nil, nil, nil, nil, func(t *testing.T, e *testEnv) {
```

Comment:
Let’s refactor doWithTestEnv to be a "do" method of the "testEnv" struct and fill the fields we use instead of passing nil as positional arguments.

Reply:
OK - upcoming to include abg's suggestions also

Reply:
I think you need a new struct i.e. testEnvOptions; testEnv is what's constructed based on those positional arguments.

Reply:
agreed - will do this